File: activity/20200904/RabbitLee/Ones_and_Zeroes.cpp
class Solution {
public:
    int findMaxForm(vector<string>& strs, int m, int n) {
        int len = strs.size();
        vector<int> zeros(len, 0), ones(len, 0);
        for (int i = 0; i < len; i++) {
            int zero = 0, one = 0;
            for (auto c : strs[i]) {
                if (c == '0') zero++;
                else one++;
            }
            zeros[i] = zero;
            ones[i] = one;
        }
        vector<vector<vector<int>>> f(len + 1, vector<vector<int>>(m + 1, vector<int>(n + 1, 0)));
        for (int i = 1; i <= len; i++) {
            for (int j = 0; j <= m; j++) {
                for (int k = 0; k <= n; k++) {
                    if (j >= zeros[i - 1] && k >= ones[i - 1] 
                        && f[i - 1][j][k] < f[i - 1][j - zeros[i - 1]][k - ones[i - 1]] + 1) {
                        f[i][j][k] = f[i - 1][j - zeros[i - 1]][k - ones[i - 1]] + 1;
                    } else {
                        f[i][j][k] = f[i - 1][j][k];
                    }
                }
            }
        }
        return f[len][m][n];
    }
};

File: activity/20200904/scau-flyingbird/OnesAndZeroes.go
func findMaxForm(strs []string, m int, n int) int {
	dp := make([][]int, m+1)
	for i := 0; i < len(dp); i++ {
		dp[i] = make([]int, n+1)
	}

	for k := 0; k < len(strs); k++ {
		sm, sn := findZeroOne(strs[k])
		for i := m; i >= sm; i-- {
			for j := n; j >= sn; j-- {
				dp[i][j] = Max(dp[i][j], dp[i-sm][j-sn]+1)
			}
		}
	}
	return dp[m][n]
}

func Max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}

func findZeroOne(str string) (m int, n int) {
	s := []byte(str)
	for i := 0; i < len(s); i++ {
		if s[i] == '0' {
			m++
		} else if s[i] == '1' {
			n++
		}
	}
	return m, n
}
